src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets

```diff
@@ -339,23 +341,37 @@ The .NET Foundation licenses this file to you under the MIT license.
 
     <MakeDir Directories="$([System.IO.Path]::GetDirectoryName($(NativeBinary)))" />
 
-    <PropertyGroup>
-      <CompileEmccArgs>-c</CompileEmccArgs>
-      <CompileEmccArgs>$(CompileEmccArgs) -s DISABLE_EXCEPTION_CATCHING=0</CompileEmccArgs>
-      <CompileEmccArgs Condition="'$(NativeDebugSymbols)' == 'true'">$(CompileEmccArgs) -g3</CompileEmccArgs>
-      <!-- TODO-LLVM: remove when we update to Emscripten with https://github.com/emscripten-core/emscripten/pull/18443 -->
-      <CompileEmccArgs>$(CompileEmccArgs) -s USE_SDL=0 $(EmccOptimizationSetting)</CompileEmccArgs>
+    <PropertyGroup Condition="'$(_targetOS)' == 'browser'">
+      <CompileWasmArgs>-c</CompileWasmArgs>
+      <CompileWasmArgs>$(CompileWasmArgs) -s DISABLE_EXCEPTION_CATCHING=0</CompileWasmArgs>
+      <CompileWasmArgs Condition="'$(NativeDebugSymbols)' == 'true'">$(CompileWasmArgs) -g3</CompileWasmArgs>
+      <!-- TODO-LLVM: remove -s USE_SDL=0 when we update to Emscripten with https://github.com/emscripten-core/emscripten/pull/18443 -->
+      <CompileWasmArgs>$(CompileWasmArgs) -s USE_SDL=0 $(EmccOptimizationSetting)</CompileWasmArgs>
 
       <ScriptExt Condition="'$(OS)' == 'Windows_NT'">.bat</ScriptExt>
-      <EmccPath>&quot;$(EMSDK)/upstream/emscripten/emcc$(ScriptExt)&quot;</EmccPath>
+      <WasmCompilerPath>&quot;$(EMSDK)/upstream/emscripten/emcc$(ScriptExt)&quot;</WasmCompilerPath>
+      <WasmLinkerPath>$(WasmCompilerPath)</WasmLinkerPath>
     </PropertyGroup>
 
-    <Exec Command="$(EmccPath) &quot;%(LlvmObjects.Identity)&quot; -o &quot;%(LlvmObjects.NativeObject)&quot; $(CompileEmccArgs)" Condition="'$(EMSDK)' != ''" />
-    <Message Text="Emscripten not found, not linking WebAssembly. To enable WebAssembly linking, install Emscripten and ensure the EMSDK environment variable points to the directory containing upstream/emscripten/emcc.bat"
-             Condition="'$(EMSDK)' == ''" />
+   <PropertyGroup Condition="'$(_targetOS)' == 'wasi'">
+      <CompileWasmArgs>-fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -disable-lsr --sysroot=&quot;$(WASI_SDK_PATH)/share/wasi-sysroot&quot; -pthread -target wasm32-wasi-threads -c</CompileWasmArgs>
+      <CompileWasmArgs Condition="'$(NativeDebugSymbols)' != 'true'">$(CompileWasmArgs) -O2</CompileWasmArgs>
+      <CompileWasmArgs Condition="'$(NativeDebugSymbols)' == 'true'">$(CompileWasmArgs) -g3</CompileWasmArgs>
+
+      <ExeExt Condition="'$(OS)' == 'Windows_NT'">.exe</ExeExt>
+      <!-- using Emscripten's clang++ because of a crash in wasi-sdk's clang++ (https://github.com/WebAssembly/wasi-sdk/issues/326) -->
+      <WasmCompilerPath>&quot;$(EMSDK)/upstream/bin/clang++$(ExeExt)&quot;</WasmCompilerPath>
+      <WasmLinkerPath>&quot;$(WASI_SDK_PATH)/bin/clang&quot;</WasmLinkerPath>
```

Comment:
Yes, it runs out of memory supposedly.  I have 64GB and in task manager it move s when compiling but not much, it could spike within task managers sample period, so I might not be seeing it.  Wasi SDK don't use tip of tree for LLVM so we aren't expecting a new version until v17.  I will raise an issue with LLVM as there are other differences with emscripten, so the bug may still be there.